### PR TITLE
Pagination module

### PR DESCRIPTION
**Pull request by StKostyantin**
This PR is implemented pagination module

### CODE REVIEW

File: modules/pagination/client-react/containers/DataProvider.jsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    constructor(props) {
+      super(props);
+      this.state = { items: null };
+    }
+
+    componentDidMount() {
+      this.loadData(0, 'replace');
+    }
+
+    loadData = (offset, dataDelivery) => {
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/DataProvider.jsx]: Logic duplication.

File: modules/pagination/client-react/containers/PaginationDemo.native.jsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      <TouchableOpacity style={styles.postWrapper}>
+        <Text style={styles.text}>{title}</Text>
+      </TouchableOpacity>
+    );
+  };
+
+  render() {
+    const { t, items } = this.props;
+    const { pagination } = this.state;
+    const options = [
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/PaginationDemo.native.jsx]: `render` is not a place for defining data.

File: modules/pagination/client-react/containers/DataProvider.jsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  const loadData = useCallback(
+    (offset, dataDelivery) => {
+      const newEdges = allEdges.slice(offset, offset + itemsNumber);
+      const edges = dataDelivery === 'add' ? (!items ? newEdges : [...items.edges, ...newEdges]) : newEdges;
+      const endCursor = edges[edges.length - 1].cursor;
+      const hasNextPage = endCursor < allEdges[allEdges.length - 1].cursor;
+      setItems({
+        totalCount: allEdges.length,
+        pageInfo: {
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/DataProvider.jsx]: Simplify like `{ endCursor: endCursor }` -> `{ endCursor }`

File: modules/pagination/client-react/containers/DataProvider.jsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+const generateEdgesArray = quantity => {
+  const arr = [];
+  for (let i = 1; i <= quantity; i++) {
+    arr.push({ cursor: i, node: { id: i, title: 'Item ' + i } });
+  }
+  return arr;
+};
+
+const itemsNumber = settings.pagination.web.itemsNumber;
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/DataProvider.jsx]: `itemsNumber` -> `limit`

File: modules/pagination/client-react/containers/DataProvider.jsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+const allEdges = generateEdgesArray(47);
+
+export const useDataProvider = () => {
+  const [items, setItems] = useState(null);
+
+  useEffect(() => {
+    loadData(0, 'replace');
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/DataProvider.jsx]: Why data delivery type (`'replace'`) is hardcoded?

File: modules/pagination/client-react/index.jsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import { NavLink, Route } from 'react-router-dom';
+import { translate } from '@restapp/i18n-client-react';
+import { MenuItem } from '@restapp/look-client-react';
+import ClientModule from '@restapp/module-client-react';
+import PaginationDemo from './containers/PaginationDemo';
+
+import resources from './locales';
+
+const NavLinkWithI18n = translate()(({ t }) => (
```
[the remark this answers is not in the document]
Comment (StKostyantin) [reply modules/pagination/client-react/index.jsx]: Fixed

File: modules/pagination/client-react/containers/PaginationDemo.native.jsx (hunk 2 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/PaginationDemo.native.jsx]: Fixed

File: modules/pagination/client-react/containers/DataProvider.jsx (hunk 3 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/DataProvider.jsx]: Fixed

File: modules/pagination/client-react/containers/DataProvider.jsx (hunk 4 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/DataProvider.jsx]: Fixed

File: modules/pagination/client-react/containers/DataProvider.jsx (hunk 5 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/DataProvider.jsx]: Fixed

File: modules/pagination/client-react/components/PaginationDemoView.native.jsx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+PaginationDemoView.propTypes = {
+  t: PropTypes.func,
+  items: PropTypes.object,
+  handlePageChange: PropTypes.func,
+  renderItem: PropTypes.func,
+  pagination: PropTypes.string
+};
+
+const styles = StyleSheet.create({
```
[the remark this answers is not in the document]
Comment (StKostyantin) [reply modules/pagination/client-react/components/PaginationDemoView.native.jsx]: Fixed

File: modules/pagination/client-react/containers/DataProvider.jsx (hunk 1 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/DataProvider.jsx]: Fixed

File: modules/pagination/client-react/index.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (evgenijolshanskij) [reply modules/pagination/client-react/index.jsx]: `t` function should be modified as well.

File: modules/pagination/client-react/components/styles.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { StyleSheet } from 'react-native';
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/components/styles.js]: I don't think `components` is the right placeholder for `styles`.

File: modules/pagination/client-react/containers/DataProvider.jsx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  return arr;
+};
+
+const {
+  pagination: { web: limit, type }
+} = settings;
+
+const allEdges = generateEdgesArray(47);
+
+const createLoadData = (offset, dataDelivery, items) => {
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/DataProvider.jsx]: `createLoadData` - what does this mean?

File: modules/pagination/client-react/containers/DataProvider.jsx [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+const generateEdgesArray = quantity => {
+  const arr = [];
+  for (let i = 1; i <= quantity; i++) {
+    arr.push({ cursor: i, node: { id: i, title: 'Item ' + i } });
+  }
+  return arr;
+};
+
+const {
+  pagination: { web: limit, type }
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/DataProvider.jsx]: What if it is needed to use different data set sizes or types on other pages? How should we handle this?

File: modules/pagination/client-react/containers/PaginationDemo.native.jsx [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+class PaginationDemo extends Component {
+  static propTypes = {
+    t: PropTypes.func,
+    items: PropTypes.object,
+    loadData: PropTypes.func
+  };
+
+  state = {
+    pagination: 'standard',
+    options: [
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/PaginationDemo.native.jsx]: Why in the state? Is it supposed to be changed?

File: modules/pagination/client-react/containers/PaginationDemo.native.jsx [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            renderItem={this.renderItem}
+            pagination={pagination}
+          />
+        )}
+      </View>
+    );
+  }
+}
+
+const styles = StyleSheet.create({
```
Comment (evgenijolshanskij) [inline modules/pagination/client-react/containers/PaginationDemo.native.jsx]: ?

File: modules/pagination/client-react/components/styles.js (hunk 8 above)
Comment (StKostyantin) [reply modules/pagination/client-react/components/styles.js]: Fixed

File: modules/pagination/client-react/containers/DataProvider.jsx (hunk 9 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/DataProvider.jsx]: Fixed

File: modules/pagination/client-react/containers/DataProvider.jsx (hunk 10 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/DataProvider.jsx]: Fixed

File: modules/pagination/client-react/containers/PaginationDemo.native.jsx (hunk 11 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/PaginationDemo.native.jsx]: Fixed

File: modules/pagination/client-react/containers/PaginationDemo.native.jsx (hunk 12 above)
Comment (StKostyantin) [reply modules/pagination/client-react/containers/PaginationDemo.native.jsx]: Fixed

**Pull request closed without merging**